### PR TITLE
fixes #20941; if branches should have the same type as the if stmt

### LINKED ISSUE #20941: `discardable` procedure with arc/orc generate unnecessary `=destroy` call and crash the program.
### Description

Example code:
```nim
type Node = ref object

proc newNode(): Node {.discardable.} =
  new(result)

proc main() =
  if false:
    newNode()

main()
```

Compile with --expandArc:
```nim
var :tmpD
if false:
  :tmpD = newNode()
  :tmpD
`=destroy`(:tmpD) # <- crash for complex object because :tmpD is not avaliable
```

Add discard:
```nim
if false:
  var :tmpD
  discard
    :tmpD = newNode()
    :tmpD
  `=destroy`(:tmpD) # <- OK
```

Add a left value
```nim
if false:
  var tmp
  tmp = newNode()
  `=destroy`(tmp) # <- OK
```

### Nim Version

Nim Compiler Version 1.7.3

### PR DESCRIPTION
**Pull request by ringabout**
fixes #20941

**It only fixes the if statement without an else branch** because that one cannot be ambiguous, and it must be a statement.

### CODE REVIEW

File: compiler/injectdestructors.nim [hunk 1]
```diff
@@ -625,7 +626,7 @@ template handleNestedTempl(n, processCall: untyped, willProduceStmt = false,
         #Condition needs to be destroyed outside of the condition/branch scope
         branch[0] = p(it[0], c, s, normal)
 
-      branch[^1] = if it[^1].typ.isEmptyType or willProduceStmt:
+      branch[^1] = if it[^1].typ.isEmptyType or willProduceStmt or withoutElse:
```
Comment (Araq) [inline compiler/injectdestructors.nim]: But `it[^1].typ.isEmptyType` should really be true. Other parts in the compiler use the same check!

Comment (ringabout) [reply compiler/injectdestructors.nim]: Let me try again with setting its type to empty in the sem phase. I failed last since I didn't change the logic of `maybeVoid`.

Comment (ringabout) [reply compiler/injectdestructors.nim]: If `it[^1].typ.isEmptyType` is true, the destructors for its return is not called. Though I could probably make an exception for implicit discardable return type.

```nim
proc main() =
  if false:
    newNode()
```

**Comment by ringabout** [comment]
Will revisit it later.
